artemis-commons/src/main/java/org/apache/activemq/artemis/utils/PasswordMaskingUtil.java

```diff
@@ -108,22 +109,34 @@ public static HashProcessor getHashProcessor() {
 
       // semi colons
       String[] parts = codecDesc.split(";");
-
       if (parts.length < 1)
          throw new ActiveMQException(ActiveMQExceptionType.ILLEGAL_STATE, "Invalid PasswordCodec value: " + codecDesc);
 
       final String codecClassName = parts[0];
 
       // load class
-      codecInstance = AccessController.doPrivileged(new PrivilegedAction<SensitiveDataCodec<String>>() {
-         @Override
-         public SensitiveDataCodec<String> run() {
-            ClassLoader loader = Thread.currentThread().getContextClassLoader();
+      codecInstance = AccessController.doPrivileged((PrivilegedAction<SensitiveDataCodec<String>>) () -> {
+         ServiceLoader<SensitiveDataCodec> serviceLoader = ServiceLoader.load(SensitiveDataCodec.class, PasswordMaskingUtil.class.getClassLoader());
+         try {
+            // Service load the codec, if a service is available
+            for (SensitiveDataCodec<String> codec : serviceLoader) {
+               if ((codec.getClass().getCanonicalName()).equals(codecClassName)) {
+                  return codec.getClass().newInstance();
+               }
+            }
+            // If a service is not available, load the codec class using this class's class loader
+            return (SensitiveDataCodec<String>) PasswordMaskingUtil.class.getClassLoader().loadClass(codecClassName).newInstance();
+         } catch (Exception e) {
             try {
-               Class<?> clazz = loader.loadClass(codecClassName);
-               return (SensitiveDataCodec<String>) clazz.newInstance();
-            } catch (Exception e) {
-               throw ActiveMQUtilBundle.BUNDLE.errorCreatingCodec(e, codecClassName);
+               // If service loading failed, load the codec class using this class's class loader
+               return (SensitiveDataCodec<String>) PasswordMaskingUtil.class.getClassLoader().loadClass(codecClassName).newInstance();
```

Comment:
I have cleaned up the code here. It should not attempt to load the codec twice using this class's class loader, if it fail had the first time. 

I also updated the documentation to mention that the custom password codec can be service loaded.